### PR TITLE
Support batched tensors in another part of `Nx.LinAlg` functions 

### PR DESCRIPTION
**Pull request by john-sonz**
Batched tensors support for:
- `Nx.LinAlg.eigh`
- `Nx.LinAlg.invert`
- `Nx.LinAlg.solve`
- `Nx.LinAlg.triangular_solve`

### CODE REVIEW

File: nx/lib/nx/lin_alg.ex [hunk 1]
```diff
@@ -954,19 +1052,14 @@ defmodule Nx.LinAlg do
...
-          {{n}, {n, n}}
-
-        shape ->
-          raise ArgumentError,
-                "eigh/2 expects a square tensor, got tensor with shape: #{inspect(shape)}"
-      end
+    {eigenvals_shape, eigenvecs_shape} = Nx.Shape.eigh(shape)
+    rank = tuple_size(shape)
+    eigenvals_name = List.duplicate(nil, rank - 1)
+    eigenvecs_name = List.duplicate(nil, rank)
```
Comment (josevalim) [inline nx/lib/nx/lin_alg.ex:1058]:
```suggestion
    eigenvecs_name = List.duplicate(nil, rank)
    eigenvals_name = tl(eigenvecs_name)
```

File: nx/lib/nx/shape.ex [hunk 2]
```diff
@@ -1795,6 +1795,34 @@ defmodule Nx.Shape do
...
+      )
+    end
+
+    [
+      [unchanged_shape, [m]],
+      [unchanged_shape, [m, m]]
+    ]
+    |> Enum.map(&List.flatten/1)
+    |> Enum.map(&List.to_tuple/1)
+    |> List.to_tuple()
```
Comment (josevalim) [inline nx/lib/nx/shape.ex:1816]:
```suggestion
    {
      List.to_tuple(unchanged_shape ++ [m]),
      List.to_tuple(unchanged_shape ++ [m, m])
    }
```

File: nx/lib/nx/shape.ex [hunk 3]
```diff
@@ -1849,10 +1917,42 @@ defmodule Nx.Shape do
...
+    cond do
+      a_batch_shape == b_1d_batch_shape and a_n == b_n ->
+        b_shape
+
+      a_batch_shape == b_2d_batch_shape and a_n == b_m ->
+        b_shape
+
+      true ->
+        expected_1d = [a_batch_shape, a_n] |> List.flatten() |> List.to_tuple()
+        expected_2d = [a_batch_shape, a_n, "m"] |> List.flatten() |> List.to_tuple()
```
Comment (josevalim) [inline nx/lib/nx/shape.ex:1942]:
```suggestion
        expected_1d = List.to_tuple(a_batch_shape ++ [a_n])
        expected_2d = List.to_tuple(a_batch_shape ++ [a_n, "m"])
```

**Comment by josevalim** [comment]
Amazing work @john-sonz! I have added some comments but they are all minor and style related. :)

File: nx/lib/nx/lin_alg.ex [hunk 4]
```diff
@@ -635,14 +640,30 @@ defmodule Nx.LinAlg do
...
+          {q_rank, _} ->
+            Nx.dot(
+              adjoint(q),
+              [q_rank - 1],
+              Enum.to_list(0..(q_rank - 3)),
+              b,
+              [q_rank - 2],
+              Enum.to_list(0..(q_rank - 3))
+            )
+        end
```
[the remark this answers is not in the document]
Comment (john-sonz) [reply nx/lib/nx/lin_alg.ex:657]: Yes! I didn't know that empty range trick, and it simplifies nicely!

**Review by josevalim — APPROVED** [review]

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**